### ISSUE TITLE
Implicit typing redeclaration error

### ISSUE THREAD

**Issue by konradha** [lfortran/lfortran#651]
When defining variables in a procedure and we enable implicit typing we get a variable redeclaration error. see #573

**Comment by Pranavchiku** [comment]
```
[SciPy]
semantic error: Symbol is already declared in the same scope
 --> /home/pranavchiku/scipy/scipy/linalg/src/lu.f:9:30
  |
9 |       integer m,n,piv(k),i,j,k,permute_l,m1,info
  |                              ^ redeclaration
  |
9 |       integer m,n,piv(k),i,j,k,permute_l,m1,info
  |                       ~ original declaration
```

**Comment by Pranavchiku** [comment]
@konradha if you could provide some directions on how to resolve this....

**Comment by certik** [comment]
Here is a minimal example, in modern Fortran:
```fortran
    subroutine f(s)
    integer :: A(s)
    integer, intent(in) :: s
    end subroutine
```
Currently we get:
```
semantic error: Variable 's' is not declared
  --> examples/expr2.f90:12:18
   |
12 |     integer :: A(s)
   |                  ^ 's' is undeclared
```

**Comment by certik** [comment]
So I think we still want to give an error message (or use implicit typing if enabled) if `s` is anything else but an argument.

If `s` is an argument, then we can just declare it in the symbol table if it is used as the length of an array, and not set the type (that will be our "mark" that `s` is being used as a length of an array), then later when we encounter `integer, intent(in) :: s`, we simply go into the symbol table and check that the type is not set, and set it correctly (to integer in this case). If there is another line `integer :: s`, then we give an error message that it is already declared.

**Comment by Pranavchiku** [comment]
Is there any provision to get ```Variable_t``` from ```symbol_t``` ?

**Comment by Pranavchiku** [comment]
How can we initialize a ```ASR::symbol_t``` with ```NULL``` as type?

**Comment by Pranavchiku** [comment]
I guess this will be fixed on merging #739
